### ISSUE TITLE
2.x: Postgres gets confused with backslash and single quotes

### ISSUE THREAD

**Issue by mfn** [cakephp/cakephp#10373]
This is a (multiple allowed):

* [x] bug

* CakePHP Version: 2.9.6
* Platform and Target: Postgres 9.6, PHP 7.1

### What you did
Insert the following value into a varchar/text field:
```PHP
$this->Model->create();
$this->Mode->save(['field' => '\\\':1']);
```

### What happened
The data was inserted but is not what I expected, upon inspection:
```PHP
var_dump($this->Model->find('all'));
```
```
array(1) {
  [0]=>
  array(1) {
    ["Model"]=>
    array(3) {
      ["id"]=>
      int(1)
      ["field"]=>
      string(1) "\'$1"
```
I.e. `field` should contain `\':1` but contains `\'$1`.

### What you expected to happen
The data should have been inserted like I provided it.

### Observation
CakePHP internally generates this query:
```SQL
INSERT INTO "public"."models" ("field") VALUES ('\'':1');
```
This is a correct Postgres SQL query.

However when I activate `log_statements = all` on the Postgres server, I see this:
`2017-03-07 16:49:54.714 UTC [30450] user@db LOG:  execute pdo_stmt_00000007: INSERT INTO "public"."models" ("field") VALUES ('\' '$1')`

Upon further debugging I can reproduce this problem with pure PHP:
```PHP
$pdo = new PDO('pgsql:host=localhost;dbname=db', 'user', 'pass');
$pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
$sql = <<<SQL
INSERT INTO models (field) values('\'':1');
SQL;
$pdo->query($sql);
```
It works correctly when using prepared statements:
```PHP
<?php
$pdo = new PDO('pgsql:host=localhost;dbname=db', 'user', 'pass');
$pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
$params = ['\\\':1'];
$sql = 'INSERT INTO posts (field) values(?)';
$query = $pdo->prepare($sql);
$query->execute($params);
```

In my tests, all DML statements were affected (insert, update, select; didn't try delete yet).

**Comment by markstory** [comment]
If you can reproduce this issue with raw PDO, then its not a CakePHP issue.

**Comment by mfn** [comment]
Reported it at https://bugs.php.net/bug.php?id=74220

**Comment by mfn** [comment]
FTR, no matter where the bug is, I needed to get CakePHP work with it.

My fix was to overwrite `\Postgres::value` with the following implementation:
```PHP
    public function value($data, $column = null, $null = true) {
        $value = parent::value($data, $column, $null);

        if (!is_string($value)) {
            return $value;
        }
        if (!isset($value{0})) {
            return $value;
        }
        if ($value{0} !== "'") {
            return $value;
        }

        return 'E' . str_replace('\\', '\\\\', $value);
    }
```
This code looks really scary, having to work with things like manually escaping, but I could not figure out any other way to get quoting reliable working.

I wrote some tests in our code base and so far couldn't break it (which doesn't mean it's bug free).

Because I already have written a slightly altered Postres driver for CakePHP because of https://github.com/cakephp/cakephp/issues/6036 , I extended it with this fix which is also available at https://github.com/mfn/cakephp2-postgres-no-meta/
